modules/apps/portal-language/portal-language-override-web/src/main/java/com/liferay/portal/language/override/web/internal/portlet/action/ImportTranslationsMVCActionCommand.java

```diff
@@ -83,33 +79,30 @@ private void _importTranslations(
 			return;
 		}
 
-		Properties languageProperties = new Properties();
-
-		languageProperties.load(
-			new InputStreamReader(
-				new FileInputStream(file), StandardCharsets.UTF_8));
-
-		if (languageProperties.size() == 0) {
-			SessionErrors.add(actionRequest, "fileInvalid");
-
-			return;
+		try {
+			_ploEntryService.importPLOEntries(
+				Files.newInputStream(file.toPath()), languageId);
 		}
+		catch (ImportTranslationsException.InvalidPropertiesFile
+					importTranslationsException) {
 
-		Enumeration<String> enumeration =
-			(Enumeration<String>)languageProperties.propertyNames();
+			SessionErrors.add(
+				actionRequest, "fileInvalid", importTranslationsException);
+		}
+		catch (ImportTranslationsException.InvalidTranslations
+					importTranslationsException) {
 
-		while (enumeration.hasMoreElements()) {
-			String key = enumeration.nextElement();
+			Map<Class<?>, Exception> exceptions =
+				importTranslationsException.getExceptions();
 
-			try {
-				_ploEntryService.addOrUpdatePLOEntry(
-					key, languageId, languageProperties.getProperty(key));
-			}
-			catch (PortalException portalException) {
+			for (Map.Entry<Class<?>, Exception> entry : exceptions.entrySet()) {
```

Comment:
I'm not sure I've seen this pattern in the Liferay codebase before - it seems sensible but Brian may not like it.

Reply:
Yep, neither did I. But, how else would we make sure that all errors thrown during the file import are going to be shown in the UI?

Reply:
Fair enough. I guess perhaps the `Map` is the strangest part - any reason not to use a `List` of exception entries since you can get the class of each one here instead?

Reply:
The idea was to not have the same exception more than once. But maybe a `Set<Exception>` solves this???

Reply:
Not sure, how are two exceptions compared, by their message only? Let me check

Reply:
but each exception is unique, right? Meaning maybe they are the same class but they have different contents that we want to display. Am I misunderstanding the intent?

Reply:
I think it would be better to have "repeats" than potentially lose information, right?

Reply:
You are right, but for this scenario we're sure we do not have exceptions of the same class with different messages. So if from the 10 keys the user has, 8 are really large, they would see

```
Key must not have more than %s characters
Key must not have more than %s characters
Key must not have more than %s characters
Key must not have more than %s characters
```

Reply:
No wait, I'm wrong

Reply:
The `SessionErrors.add` is actually a put, if it exists, it replaces it

Reply:
I'll change it to a list

Reply:
@drewbrokke Just finished it, let me know if something is missing :)